### PR TITLE
Update user fix

### LINKED ISSUE #544: error updating users via admin panel
see [thread on coinbistro](https://coinbistro.com/threads/oSdGr1vyQkaUTI02fi76Eg/posts?start=1#N2SeUSDOQmqtTmhMrJG1dg)

when modifying a user via the admin panel, clicking the save button results in an error `There was an error updating user [username]`

### PR DESCRIPTION
**Pull request by akinsey**
Fixes #544

### CODE REVIEW

File: modules/ept-users/authorization/update.js [hunk 1]
```diff
@@ -105,33 +123,53 @@ function isUsernameUnique(server, userId, username) {
 }
 
 // old password valid
-function isPasswordValid(server, userId, password) {
-  if (password) {
+function isPasswordValid(server, userId, oldPassword, newPassword) {
```
Comment (unenglishable) [inline modules/ept-users/authorization/update.js]: it appears that this function is not used anywhere

**Comment by unenglishable** [comment]
check these noted changes:

* authorization for admin update users works (does not return error code)

* authorization for user update

* ui displays good error messages

* users with higher priority or admin/mod status can modify other users' properties

* users with lower priority or admin/mod status cannot modify other users' properties

File: modules/ept-users/authorization/update.js [hunk 2]
```diff
@@ -105,33 +123,53 @@ function isUsernameUnique(server, userId, username) {
...
+    var isAdmin = server.plugins.acls.getACLValue(auth, 'users.update.bypass.priority.admin');
+    if (paramUser.email === payload.email) { return true; }
+    else if (isAdmin) { return true; }
+    else { return Promise.reject(Boom.badRequest('Cannot change email')); }
+  })
+  .error(() => { return Promise.reject(Boom.badRequest()); })
+}
+
 // reject passwords change attempts for other users
 function rejectPassword(payload) {
```
Comment (unenglishable) [inline modules/ept-users/authorization/update.js]: it appears that this function is not used anywhere

Comment (unenglishable) [reply modules/ept-users/authorization/update.js]: `rejectPassword`

File: modules/ept-users/authorization/update.js [hunk 3]
```diff
@@ -156,6 +194,6 @@ function hasPriority(server, auth, paramsId, authedId) {
     if (same && aid <= pid) { return true; }
     // current has higher priority than referenced
     else if (lower && aid < pid) { return true; }
-    else { return Promise.reject(Boom.badRequest()); }
+    else { return Promise.reject(Boom.badRequest('This user has the same or higher priveleges than you, you may not edit their account')); }
```
Comment (unenglishable) [inline modules/ept-users/authorization/update.js]: ** privileges

File: modules/ept-users/authorization/update.js [hunk 4]
```diff
@@ -105,33 +123,53 @@ function isUsernameUnique(server, userId, username) {
...
+      type: 'validatePassword',
+      server: server,
+      userId: userId,
+      password: password
+    });
   }
   else { return true; }
 }
 
+
```
Comment (unenglishable) [inline modules/ept-users/authorization/update.js]: extra newline

File: modules/ept-users/authorization/update.js [hunk 5]
```diff
@@ -105,33 +123,53 @@ function isUsernameUnique(server, userId, username) {
...
+// remove email from payload for other users
+function rejectEmail(server, auth, payload, paramsId) {
+  return server.db.users.find(paramsId)
+  .then(function(paramUser) {
+    var isAdmin = server.plugins.acls.getACLValue(auth, 'users.update.bypass.priority.admin');
+    if (paramUser.email === payload.email) { return true; }
+    else if (isAdmin) { return true; }
+    else { return Promise.reject(Boom.badRequest('Cannot change email')); }
+  })
+  .error(() => { return Promise.reject(Boom.badRequest()); })
```
Comment (unenglishable) [inline modules/ept-users/authorization/update.js]: should we include the error message in this reject?

**Comment by akinsey** [comment]
Yes, just supply e to the function and pass e to bad request. 

I’ve looked at your other issues as well, everything you brought up is valid, so feel free to correct the issues if you have time, if not I’ll do it on Monday. Thanks 🙏

**Comment by unenglishable** [comment]
password rules don't apply when using the `edit password` interface.

for example, if i used old password:  `password` and did new password and confirmation:  `p`, the interface shows green checkmarks, ignoring the 8 character minimum.  an error is returned from the api, but it does not indicate what went wrong.

if possible, we should re-use the same interface that's available when signing up for the forum.  that would ensure that both creating and updating password behave the same.

File: modules/ept-users/authorization/update.js [hunk 6]
```diff
@@ -5,57 +5,75 @@ var querystring = require('querystring');
...
 function isAllowedToChangeUsername(server, auth, userId, username) {
   // get referenced user's username
   return server.db.users.find(userId)
   .then(function(paramUser) {
-    if (paramUser.username == username) { return true }
+    if (paramUser.username === username) { return true; }
     else {
       return server.authorization.build({
-        error: Boom.forbidden(),
+        error: Boom.forbidden('You do not have the appropriate permissions to change other user\'s usernames'),
```
Comment (unenglishable) [inline modules/ept-users/authorization/update.js]: ** `this user's`
OR
** `other users'`

**Comment by unenglishable** [comment]
updating website shows errors correctly in the admin panel, but not from the user's profile page

**Comment by unenglishable** [comment]
getting client error

```ReferenceError: canUpdate is not defined
    at controller.canUpdateUsername (webpack:///./app/components/profile/profile.directive.js?:59)```

File: modules/ept-users/authorization/update.js (hunk 6 above)
Comment (unenglishable) [reply modules/ept-users/authorization/update.js]: fixed in #555

File: modules/ept-users/authorization/update.js (hunk 4 above)
Comment (unenglishable) [reply modules/ept-users/authorization/update.js]: fixed in #555

File: modules/ept-users/authorization/update.js (hunk 5 above)
Comment (unenglishable) [reply modules/ept-users/authorization/update.js]: fixed in #555

File: modules/ept-users/authorization/update.js (hunk 2 above)
Comment (unenglishable) [reply modules/ept-users/authorization/update.js]: fixed in #555

File: modules/ept-users/authorization/update.js (hunk 3 above)
Comment (unenglishable) [reply modules/ept-users/authorization/update.js]: fixed in #555

**Pull request merged**